pkg/cloud/cloud_test.go

```diff
@@ -553,6 +597,101 @@ func executeDescribeSnapshotsTest(t *testing.T, c *cloud, snapshotIDs, snapshotN
 	}
 }
 
+func TestCheckDesiredState(t *testing.T) {
+	testCases := []struct {
+		name           string
+		volumeId       string
+		desiredSizeGiB int32
+		options        *ModifyDiskOptions
+		expErr         error
+	}{
+		{
+			name:           "sucess: normal path",
+			volumeId:       "vol-001",
+			desiredSizeGiB: 5,
+			options: &ModifyDiskOptions{
+				VolumeType: VolumeTypeGP2,
+				IOPS:       3000,
+				Throughput: 1000,
+			},
+		},
+		{
+			name:           "failure: volume is still being expanded",
+			volumeId:       "vol-001",
+			desiredSizeGiB: 500,
+			options: &ModifyDiskOptions{
+				VolumeType: VolumeTypeGP2,
+				IOPS:       3000,
+				Throughput: 1000,
+			},
+			expErr: fmt.Errorf("volume \"vol-001\" is still being expanded to 500 size"),
```

Comment:
This hardcoding of error message in test seems like a similar code smell to me as mentioned in our backlog grooming (e.g. we are string matching on error message).

Can we wrap an error in the [modification code](https://github.com/kubernetes-sigs/aws-ebs-csi-driver/blob/aeb6a026f0f509b2fd47b0b7aea0b9649930eafe/pkg/cloud/cloud.go#L1701) similar to [what we do for attachErr](https://github.com/kubernetes-sigs/aws-ebs-csi-driver/blob/aeb6a026f0f509b2fd47b0b7aea0b9649930eafe/pkg/cloud/cloud.go#L955), and then check that with [errors.is()](https://pkg.go.dev/errors#Is) in unit test?

Reply:
What error would we wrap? This is the original error source.

Reply:
The original source hardcodes custom errors instead of wrapping a ModifyExistingOperationErr as we do for things like attachErr. 

At the source we should create a ModifyExistingOperation err, and wrap that with any variable text (volume-id and parameter size)

Then in the test check for that error, instead of this brittle hardcoding of volume id and exact string matching.

Reply:
I don't understand why that method is "less brittle" than the current, and it causes the test to be less accurate.

Reply:
Brittleness here means the unit test would fail if we change something not as important, like the exact wording of the error message.

A less brittle test would make sure a "ModificationInProgress" err is returned here, but not need the error message to match an exact string.  

From one of my favorite resources, [Google Testing Blog: How I Learned To Stop Writing Brittle Tests](https://testing.googleblog.com/2024/04/how-i-learned-to-stop-writing-brittle.html):

"A [“brittle” test](https://abseil.io/resources/swe-book/html/ch12.html#preventing_brittle_tests) is one that fails not for real problems that would break in production, but because the test itself is fragile for innocuous reasons. Error messages, changing the order of metadata headers in a web request, or the order of calls to a heavily-mocked dependency can often cause a brittle test to fail."

Reply:
Changing the error message is breakage though, because the error message is user facing.

For example, if the function suddenly started returning the error `volume \"(nil)\" is still being expanded to -1 size` we would want to know that and fail the test.

Reply:
If we want to catch "user facing err missing volume-id", what if we have the best of both worlds with a test util function like:

func ValidateErr(testErr, expectedErr errors.error, otherExpectedStrings... string) {
    if !errors.Is(testErr, expectedErr) {
        return false
    }
    for _, s := range otherExpectedStrings {
        if !strings.Contains(testErr, s) { return false }
    }
    return true
}

Reply:
The `errors.Is` there appears to be circular - it's checking the error is a `ModificationInProgress` error, but the only reason that error type was added is for the test. The fact it is a `ModificationInProgress` error does not matter and will not affect any other code - it's a requirement added purely by the test.

(That would actually be a great example of a brittle test, because someone that notices that `ModificationInProgress` is unused and refactors it out of the code will suddenly be faced with a failing test)

Reply:
Gotcha, would it be helpful to rephrase my desires that the unit test be verifying:

- If volume modification is already in progress, confirm SOME error is returned.
  - Because we want to protect established contract in the least specific way
- Error message has volume-id + relevant mutable parameter somewhere
  - Because it will flag us accidentally screwing up customer facing message

I'm fine with not having a match on `ModificationInProgress`, I just wanted to reduce unit test excessive verification  from "Exact error" to "shared error like attachErr`. I think your point of "let's make sure volume-id is in msg" is an even better way to reduce that excessive verification to just something that matters to customer. 

Would you be aligned with ^^ @ConnorJC3?

Reply:
Offline thread occurred, it seems like ^^ would not be best option.

I'm resolving this thread and will bring up a clearer version of my thoughts on verification bounds in unit tests through our next team retro!

Thanks @ElijahQuinones @ConnorJC3 for the learnings.